### PR TITLE
DOC: Clarify output of diff (returned type and possible overflow)

### LINKED ISSUE #28909: DOC: diff() when using unsigned ints
#### Code Sample, a copy-pastable example if possible

```python
>>> import numpy as np
>>> import pandas as pd
>>> 
>>> s = pd.Series([1, 0], dtype=np.uint8)
>>> s.diff()
0      NaN
1    255.0
dtype: float64
```
#### Problem description

The current version of the docs don't reflect on what happens when a Series or DataFrame containing unsigned ints are used with the `diff()` function, especially for cases where integer overflow occurs.

NumPy handles this in their docs for `numpy.diff()`, stating:

> For unsigned integer arrays, the results will also be unsigned. This should not be surprising, as the result is consistent with calculating the difference directly:
```python
>>> u8_arr = np.array([1, 0], dtype=np.uint8)
>>> np.diff(u8_arr)
array([255], dtype=uint8)
>>> u8_arr[1,...] - u8_arr[0,...]
255
```

The behaviour of `pandas.Series.diff()`  and `pandas.DataFrame.diff()` during calculation is the same as the NumPy equivalent, but the actual return type is no longer an unsigned integer; it will always return with a `float64` type.

#### Expected Output

The current output _can_ be expected when the user is perfectly aware that integer overflow might possibly occur during the calculation of the `diff()` output, even though it returns as a `float64` dtype. I do think it is desired to mention this in docs similarly to the NumPy equivalent though.

However, since the output dtype of `diff()` will always be `float64`, this also allows to handle the case differently: Instead of storing the output of the calculation:
https://github.com/pandas-dev/pandas/blob/0c4404b5b85c085e5b6cd26be42e81e5a197fd8d/pandas/core/algorithms.py#L1964
into a float64 dtyped array, it would also be possible to process the values as float instead of integers? This would generate the output:
```
0    NaN
1   -1.0
dtype: float64
```

I understand that this might go against the absolute difference of the elements as they are defined inside of the Series or DataFrame, but it would better fit the output type? Maybe just a warning for the overflow is the best middle ground instead? Would love to hear your thoughts on the case...

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : None.None

pandas           : 0.25.1
numpy            : 1.17.2
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.3
setuptools       : 41.2.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by mproszewska**
- [x] closes #28909
- [x] tests passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] note about dtype in `diff` in Dataframa and Series

### CODE REVIEW

**Comment by WillAyd** [comment]
Thanks for the PR. I don't understand the comments though - can you maybe provide code for what you are trying to express?

**Comment by datapythonista** [comment]
Thanks for working on this @mproszewska. Agree with @WillAyd, I don't understand what's the comment trying to explain. I can read what's the problem in the issue, but I don't think the added comment is helping users understand. Can you give it another try?

File: pandas/core/frame.py
Comment (datapythonista) [inline pandas/core/frame.py]: My understanding from the issue, is that the main problem is the overflow, not the type.

I think adding this example from the issue to the examples, with a comment explaining why the results are counter-intuitive (the value and the type), could probably be more useful:
```python
>>> import numpy as np
>>> import pandas as pd
>>> 
>>> s = pd.Series([1, 0], dtype=np.uint8)
>>> s.diff()
0      NaN
1    255.0
dtype: float64
```

**Review by datapythonista — COMMENTED** [review]
Thanks for working on this @mproszewska. Looking better, but just one idea about how to approach the problem.

And also, I think this PR is a good opportunity to unify the two docstrings that you're updating, so there is just one, and the other method reuses it. We have a `@doc` decorator for it. The general idea would be something like:
```python
class Series:
    @doc(klass='Series', extra_params='', ...)
    def diff(...):
    """
    First discrete difference of element.

    Calculates the difference of a {klass} element compared with another
    ...
    """
    ...

class DataFrame:
    @doc(Series.diff, klass='DataFrame', extra_param="axis : {0 or 'index', 1 or 'columns'}, default 0...", ...)
    def diff(...):
    ...  # no docstring here, since it's reused the one from `Series.diff`
```

You have an example of using `@doc` in #32787

Thanks for helping with this, let me know if you have any question.

File: pandas/core/frame.py
Comment (mproszewska) [reply pandas/core/frame.py]: done

**Comment by datapythonista** [comment]
Thanks @mproszewska. Can you use another variable `{examples}` for the examples, instead of using `Appender` (we want to deprecate `Appender` and use `doc` instead).

Also, you've got conflicts. And if you can run `scripts/validate_docstrings.py pandas.Series.diff` and same for the other docstring, to make sure that they render as you expect, and that no error is reported, that would be great.

Thanks!

**Comment by mproszewska** [comment]
> Also, you've got conflicts. And if you can run `scripts/validate_docstrings.py pandas.Series.diff` and same for the other docstring, to make sure that they render as you expect, and that no error is reported, that would be great.

I can't run this script. I have numpydoc installed but I keep getting
```
Traceback (most recent call last):
  File "scripts/validate_docstrings.py", line 50, in <module>
    from numpydoc.validate import validate, Docstring  # noqa: E402 isort:skip
ModuleNotFoundError: No module named 'numpydoc.validate'
```

I checked changed docs with `help(pd.Series.diff)` and they seemed fine.

**Comment by datapythonista** [comment]
You need to install the development version of numpydoc. It's in `environment.yml`. If you're using conda and installing things in the recommended way, a simple `conda env update` in the root directory of pandas should get you the right version.

File: pandas/core/frame.py
Comment (datapythonista) [inline pandas/core/frame.py:6471]: that's unreadable, use `"""` and make the `\n` actual line breaks, so this can be read.

**Review by datapythonista — COMMENTED** [review]
Looks good, thanks for working on this @mproszewska 

Only thing is that in the examples parameter I think we usually use `textwrap.dedent`, to have the text indented, and make clear what it's the parameter, and where the function is defined. Can you have a look at other cases, and see if this is correct, please? Thanks!

**Comment by datapythonista** [comment]
Thanks @mproszewska 

Can you run `./scripts/validate_docstrings.py pandas.DataFrame.diff` and make sure it doesn't report any error please?

And you'll have to run `black` to fix the CI. Thanks!

**Comment by mproszewska** [comment]
`validate_docstrings` does not report any errors.

**Review by datapythonista — APPROVED** [review]
Thanks @mproszewska, very nice PR, lgtm

**Comment by jbrockmendel** [comment]
@mproszewska can you see if you can get the CI passing

**Comment by mproszewska** [comment]
`test_diff` was checking stacklevel of warning produced by `diff`. After my changes in documentation warning was raised in `_decorators.py` file instead of `test_algos.py`. I changed test so that stacklevel is not checked. I'm not sure if that's the best idea, but I couldn't find any other way.

File: pandas/tests/arrays/categorical/test_algos.py
Comment (datapythonista) [inline pandas/tests/arrays/categorical/test_algos.py:97]: You will have to restore those checks, they are important.

When a warning is raised, we want to raise them in the line that the user wrote, like the `Series` creation. This is what they are checking. If you run in your branch the code `s = pd.Series([1, 2, 3], dtype="category")` you'll see that this is not true anymore, and the warning will point to some internal code.

This is caused by using the doc decorator, and should be changes when we generate that `FutureWarning`. There we tell the stacklevel, the number of calls we want to go "up" to generate the warning there.

We're changing the doc decorator, so we'll have to wait until this can be merged, see: https://github.com/pandas-dev/pandas/pull/33160#pullrequestreview-400797773

**Review by datapythonista — CHANGES_REQUESTED** [review]
We need to wait for #33160 before we can merge this. But you can find where the `FutureWarning` and change the stack level for now, and may be after that PR is merged we'll need to change that number again, but that should be easy once you found where it's being defined.

**Comment by mproszewska** [comment]
I changed stacklevel. Everything should be fine now.

**Comment by jreback** [comment]
@mproszewska can you merge master and ping on green.

@datapythonista over to you.

File: pandas/core/algorithms.py
Comment (datapythonista) [inline pandas/core/algorithms.py:1860]: Can you change the value directly in line 1817? From 3 to 4.

Comment (mproszewska) [reply pandas/core/algorithms.py]: After merging with master, tests pass even with stacklevel=3, so I changed it.

File: pandas/core/sorting.py
Comment (datapythonista) [inline pandas/core/sorting.py:388]: This change doesn't seem related. Is it coming from manually fixing a commit?

Comment (mproszewska) [reply pandas/core/sorting.py]: yes, sorry for that

**Comment by datapythonista** [comment]
Thanks @mproszewska, great PR.

Can you have a look at https://pandas.pydata.org/docs/dev/reference/api/pandas.DataFrame.diff.html when it's updated, and see that everything looks as expected please?

**Pull request merged**
